### PR TITLE
Region WPS

### PR DESCRIPTION
**Pull request by rdengate**
This allows region as a WPS input parameter, and also changes the widget so that the region is selected on the map rather than on a preview map in the data catalog (this change affects Analytics as well). 

Fixes some bugs as well. For example, countries can now be selected by clicking on the map rather than only via autocomplete, and autocomplete now works if user enters a capital letter as the first letter of the region.

### CODE REVIEW

File: Documentation/Customizing/WPS-parameters-guide.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+## Guide to WPS parameters: which UI elements they map to
```
Comment (kring) [inline Documentation/Customizing/WPS-parameters-guide.md:1]: Nice!

File: lib/Models/RegionParameter.js [hunk 2]
```diff
@@ -96,7 +96,11 @@ RegionParameter.prototype.findRegionByID = function(regionID) {
     }
 
     return regionProvider.regions.filter(function(region) {
-        return region.id === regionID;
+        if (typeof region.id === 'string' && typeof regionID === 'string') {
```
Comment (kring) [inline lib/Models/RegionParameter.js]: Why is this change necessary?  I think it's ok for region IDs to be case sensitive.  Or, put another way, the replacements system should take care of any adjustments to the ID that are necessary.

File: lib/Models/WebProcessingServiceCatalogFunction.js [hunk 3]
```diff
@@ -95,6 +99,10 @@ WebProcessingServiceCatalogFunction.parameterConverters = [
                 return undefined;
             }
 
+            if (input.Identifier === "regionType") {
```
Comment (kring) [inline lib/Models/WebProcessingServiceCatalogFunction.js]: It seems a bit strange to check this here.  Can we handle it just by ordering the converters differently?  i.e. put the `regionType` converter first in the list, and then this converter works on any `LiteralData` that isn't handled by a prior converter?  That sort of thing will be necessary if someone wants to extend the set up parameters from outside TerriaJS, because they won't be able to add an actual `Identifier` check to our existing converters easily.

File: lib/Models/WebProcessingServiceCatalogFunction.js [hunk 4]
```diff
@@ -139,6 +147,44 @@ WebProcessingServiceCatalogFunction.parameterConverters = [
         }
     },
     {
+        id: 'RegionType',
+        inputToFunctionParameter: function(catalogFunction, input) {
+            if (!defined(input.LiteralData)) {
+                return undefined;
+            }
+
+            if (input.Identifier !== "regionType") {
```
Comment (kring) [inline lib/Models/WebProcessingServiceCatalogFunction.js]: `Identifier` is basically the name of the parameter, right?  What if we want to select two region types?  That's not purely theoretical, the Spatial Detailing service does exactly that.

File: lib/Models/WebProcessingServiceCatalogFunction.js [hunk 5]
```diff
@@ -139,6 +147,44 @@ WebProcessingServiceCatalogFunction.parameterConverters = [
...
+                return undefined;
+            }
+
+            var allowedValues = input.LiteralData.AllowedValues;
+            var validRegionTypes;
+            if (defined(allowedValues)) {
+                var allowed = allowedValues.Value.slice();
+                if (typeof(allowed) === "string") {
```
Comment (kring) [inline lib/Models/WebProcessingServiceCatalogFunction.js]: No parenthesis around `(allowed)`, use single quotes.

File: lib/Models/WebProcessingServiceCatalogFunction.js [hunk 6]
```diff
@@ -177,6 +223,60 @@ WebProcessingServiceCatalogFunction.parameterConverters = [
...
+             var schema = input.ComplexData.Default.Format.Schema;
+             if (schema.indexOf('http://geojson.org/geojson-spec.html#') !== 0) {
+                 return undefined;
+             }
+
+             if (schema.substring(schema.lastIndexOf('#') + 1) !== 'multipolygon') {
+                 return undefined;
+             }
+
+             if (input.Identifier.indexOf('region') === -1) {
```
Comment (kring) [inline lib/Models/WebProcessingServiceCatalogFunction.js]: I think this test is pretty suspect, too.

It seems to me, anytime a WPS wants a polygon, the UI should allow the user to specify that polygon in a number of ways, including:
- By clicking out polygon points.
- By drawing a bounding box.
- By selecting a region.
- Maybe: by selecting an existing polygon on the map?

It's up to the user how they define their polygon, it's not necessarily specified by the server (though perhaps there could be a mechanism for the server to hint at a good default).

Is that totally crazy?

File: lib/Models/WebProcessingServiceCatalogFunction.js [hunk 7]
```diff
@@ -177,6 +223,60 @@ WebProcessingServiceCatalogFunction.parameterConverters = [
...
+                 return undefined;
+             }
+
+             if (!defined(catalogFunction.regionTypeParameter)) {
+                 throw new DeveloperError('No regionTypeParameter found, but regionGeometry requires a regionTypeParameter.');
```
Comment (kring) [inline lib/Models/WebProcessingServiceCatalogFunction.js]: Why does it need a region type parameter?

File: lib/Models/WebProcessingServiceCatalogFunction.js [hunk 8]
```diff
@@ -177,6 +223,60 @@ WebProcessingServiceCatalogFunction.parameterConverters = [
...
+                 isRequired: (input.minOccurs | 0) > 0,
+                 regionProvider: catalogFunction.regionTypeParameter
+             });
+         },
+         functionParameterToInput: function(catalogFunction, parameter, value) {
+             if (!defined(value) || value === '' || !defined(value.id)) {
+                 return undefined;
+             }
+
+             // Value is an object with id and fid. We want an actual polygon for WPS, but not for analytics.
```
Comment (kring) [inline lib/Models/WebProcessingServiceCatalogFunction.js]: I'd argue that's because the input is a polygon, and we're only specifying the polygon by selecting a region as a convenience.  It's possible a different kind of input (even if it's one we don't have yet) could accept a region type and identifier.

File: lib/Models/WhyAmISpecialFunction.js [hunk 9]
```diff
@@ -29,6 +29,8 @@ var WhyAmISpecialFunction = function(terria) {
         name: 'Region Type',
         description: 'The type of region to analyze.'
     });
+    // Let the RegionParameter handle showing it in the UI, rather than displaying it directly.
+    this._regionTypeParameter.showInUi = false;
```
Comment (kring) [inline lib/Models/WhyAmISpecialFunction.js]: We can probably do this automatically at the UI level.  Look for a `RegionParameter` with a `regionType` property equal to this `RegionTypeParameter`.  If we find one, don't show the UI for this `RegionTypeParameter`.

File: lib/ReactViews/Analytics/RegionDataParameterEditor.jsx [hunk 10]
```diff
@@ -166,12 +166,13 @@ const RegionDataParameterEditor = React.createClass({
             );
         }
         return (
+            // Don't break the lines around the link to csv-geo-au, or whitespace stripping will ruin the formatting in
+            // the rendered version.
```
Comment (kring) [inline lib/ReactViews/Analytics/RegionDataParameterEditor.jsx:170]: Ugh, thanks React.

File: lib/Models/MapInteractionMode.js [hunk 11]
```diff
@@ -24,6 +24,12 @@ function MapInteractionMode(options) {
     this.onCancel = options.onCancel;
 
     /**
+     * Opportunity to display extra custom UI elements.
+     * @type {Function}
+     */
+    this.customUi = defaultValue(options.customUi, function() { return ""; });
```
Comment (kring) [inline lib/Models/MapInteractionMode.js]: This is pretty interesting.  From the standpoint of the `MapInteractionMode`, this is kind of like an opaque object.  In our React-based UI, it's a function that returns a React component.  In our old Knockout based UI, it might be a view-model or something like that, so not a function at all.  It might be good to document it this way.  Something like:

```
/**
 * Gets or sets the details of a custom user interface for this map interaction mode.  This property is not
 * used by the `MapInteractionMode` itself, so it can be anything that is suitable for the user interface.  In
 * the standard React-based user interface included with TerriaJS, this property is a function that is called
 * with no parameters and is expected to return a React component.
 * @type {Any}
 */
```

Also the default value should probably be undefined.

**Review by kring — CHANGES_REQUESTED** [review]
I was skeptical at first, but I really like the new region picking UI!

File: lib/Models/UserDrawing.js [hunk 12]
```diff
@@ -157,7 +157,9 @@ var UserDrawing = function(options) {
     // helper for dragging points around
     var that = this;
     this.dragHelper = new DragPoints(options.terria, function(customDataSource) {
-        that.onPointMoved(customDataSource);
+        if (typeof that.onPointMoved === "function") {
```
[the remark this answers is not in the document]
Comment (rdengate) [reply lib/Models/UserDrawing.js]: Sure, will change. From what I understand, javascript allows both interchangeably though... is there a style guide I should be following?

File: lib/Models/RegionParameter.js (hunk 2 above)
Comment (rdengate) [reply lib/Models/RegionParameter.js]: This is the change that allows countries to be selected by clicking on them. Otherwise you get to here and (for example) Australia is compared to australia and the region id is not returned. Do I need to fix this in regionProvider instead though?

File: lib/Models/WebProcessingServiceCatalogFunction.js (hunk 5 above)
Comment (rdengate) [reply lib/Models/WebProcessingServiceCatalogFunction.js]: thanks, changed.

File: lib/Models/WebProcessingServiceCatalogFunction.js (hunk 7 above)
Comment (rdengate) [reply lib/Models/WebProcessingServiceCatalogFunction.js]: RegionParameter takes one, which it uses to determine region ID.

File: lib/Models/MapInteractionMode.js (hunk 11 above)
Comment (rdengate) [reply lib/Models/MapInteractionMode.js]: Thanks, changed.

File: lib/Models/WhyAmISpecialFunction.js (hunk 9 above)
Comment (rdengate) [reply lib/Models/WhyAmISpecialFunction.js]: Done, thanks.

**Comment by rdengate** [comment]
Thanks @kring. I believe any comments I haven't answered above have been addressed in the offline discussion but let me know if I missed anything. This is ready for round 2 :-)

**Review by kring — CHANGES_REQUESTED** [review]
Looks good other than these two small comments.

File: lib/Models/MapInteractionMode.js [hunk 13]
```diff
@@ -24,6 +24,14 @@ function MapInteractionMode(options) {
     this.onCancel = options.onCancel;
 
     /**
+     * Gets or sets the details of a custom user interface for this map interaction mode. This property is not used by
+     * the `MapInteractionMode` itself, so it can be anything that is suitable for the user interface. In the standard
+     * React-based user interface included with TerriaJS, this property is a function that is called with no parameters
+     * and is expected to return a React component.  @type {Any}
+     */
+    this.customUi = defaultValue(options.customUi, function() { return undefined; });
```
[the remark this answers is not in the document]
Comment (rdengate) [reply lib/Models/MapInteractionMode.js]: Thanks, changed.

File: lib/Models/RegionParameter.js (hunk 2 above)
Comment (rdengate) [reply lib/Models/RegionParameter.js]: Argh! This used to fix the problem described, but you're right, that now works without my change. I've removed my change. Note that it's still comparing Australia to australia and returning no match... but the country is still selected, so I guess there's something that has changed elsewhere.

**Review by kring — APPROVED** [review]

File: lib/Models/RegionParameter.js (hunk 2 above)
Comment (rdengate) [reply lib/Models/RegionParameter.js]: Is this with region type COUNTRY?
